airflow/io/path.py

```diff
@@ -129,7 +130,11 @@ def __new__(
 
             return _cls(_cls._format_parsed_parts(drv, root, parts, **other_kwargs), **new_kwargs)
 
-        url = stringify_path(other)
+        if isinstance(other, Dataset):
+            url = other.uri
+        else:
+            url = stringify_path(other)
```

Comment:
I wonder if we should do this from the other direction, making Dataset a PathLike instead by implementing `__fspath__` and compatible with `stringify_path`.

Reply:
Mmm Iike that too. And maybe both ways then?